1. What is a neural network? What are the general steps required to build a neural network? 

It is a network with extreme complex interactions. Consists of nodes, input layer, hidden layer and output layers. 
Nodes contain information from input data and these make the model capture more interactions. 
Steps required are:
1. Specify Architecture
2. Compile - specify optimizer and loss function
3. Model Fitting
4. Predict

2. Generally, how do you check the performance of a neural network? Why? 

To achieve the best performance of a model, we need to tune the hyperparameters. 
1. Learning algorithm - controls the adjustment of weights with respect to loss function
2. Activation Function - this by which the model process and pass the information. Ex: binary, sigmoid, tanh, softmax, ReLU, Leaky ReLU
3. Gradient - batch gradient descent and stochastic Grdient descent
4. Epochs - number of time the model goes through the entire training set
5. Regularization - by penalizing loss function
6. Optimization - updates the weights parameters to minimize the loss function. Ex: Adam, RMSProp, Mometum, Adagrad, Adadelta


3. Create a neural network using keras to predict the outcome of either of these datasets:
Cardiac Arrhythmia: https://archive.ics.uci.edu/ml/datasets/Arrhythmia 
Abalone age: https://archive.ics.uci.edu/ml/datasets/Abalone

In [108]:
import numpy as np
import pandas as pd

In [109]:
#Loaded csv data from local
df_new = pd.read_csv('/Users/trimpu/Downloads/abalone.data')
#converted to a DataFrame
df = pd.DataFrame(df_new)
df

M  0.455  0.365  0.095   0.514  0.2245   0.101    0.15  15
0     M  0.350  0.265  0.090  0.2255  0.0995  0.0485  0.0700   7
1     F  0.530  0.420  0.135  0.6770  0.2565  0.1415  0.2100   9
2     M  0.440  0.365  0.125  0.5160  0.2155  0.1140  0.1550  10
3     I  0.330  0.255  0.080  0.2050  0.0895  0.0395  0.0550   7
4     I  0.425  0.300  0.095  0.3515  0.1410  0.0775  0.1200   8
...  ..    ...    ...    ...     ...     ...     ...     ...  ..
4171  F  0.565  0.450  0.165  0.8870  0.3700  0.2390  0.2490  11
4172  M  0.590  0.440  0.135  0.9660  0.4390  0.2145  0.2605  10
4173  M  0.600  0.475  0.205  1.1760  0.5255  0.2875  0.3080   9
4174  F  0.625  0.485  0.150  1.0945  0.5310  0.2610  0.2960  10
4175  M  0.710  0.555  0.195  1.9485  0.9455  0.3765  0.4950  12

[4176 rows x 9 columns]

In [110]:
df.columns = ['Sex', 'Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weigth', 'Viscera weight', 'Shell weigth', 'Rings']
df.head(2)

Sex  Length  Diameter  Height  Whole weight  Shucked weigth  Viscera weight  \
0   M    0.35     0.265   0.090        0.2255          0.0995          0.0485   
1   F    0.53     0.420   0.135        0.6770          0.2565          0.1415   

   Shell weigth  Rings  
0          0.07      7  
1          0.21      9

In [111]:
#Preprocess your dataset. Indicate which steps worked and which didn’t. Include your thoughts on why certain steps worked and certain steps didn’t. 
df.isnull().sum()

Sex               0
Length            0
Diameter          0
Height            0
Whole weight      0
Shucked weigth    0
Viscera weight    0
Shell weigth      0
Rings             0
dtype: int64

In [112]:
df.corr()

Length  Diameter    Height  Whole weight  Shucked weigth  \
Length          1.000000  0.986813  0.827552      0.925255        0.897905   
Diameter        0.986813  1.000000  0.833705      0.925452        0.893159   
Height          0.827552  0.833705  1.000000      0.819209        0.774957   
Whole weight    0.925255  0.925452  0.819209      1.000000        0.969403   
Shucked weigth  0.897905  0.893159  0.774957      0.969403        1.000000   
Viscera weight  0.903010  0.899726  0.798293      0.966372        0.931956   
Shell weigth    0.897697  0.905328  0.817326      0.955351        0.882606   
Rings           0.557123  0.575005  0.558109      0.540818        0.421256   

                Viscera weight  Shell weigth     Rings  
Length                0.903010      0.897697  0.557123  
Diameter              0.899726      0.905328  0.575005  
Height                0.798293      0.817326  0.558109  
Whole weight          0.966372      0.955351  0.540818  
Shucked weigth        0.931956      0.882606  0.421256  
Viscera weight        1.000000      0.907647  0.504274  
Shell weigth          0.907647      1.000000  0.628031  
Rings                 0.504274      0.628031  1.000000

In [113]:
df.describe()

Length     Diameter       Height  Whole weight  Shucked weigth  \
count  4176.000000  4176.000000  4176.000000   4176.000000      4176.00000   
mean      0.524009     0.407892     0.139527      0.828818         0.35940   
std       0.120103     0.099250     0.041826      0.490424         0.22198   
min       0.075000     0.055000     0.000000      0.002000         0.00100   
25%       0.450000     0.350000     0.115000      0.441500         0.18600   
50%       0.545000     0.425000     0.140000      0.799750         0.33600   
75%       0.615000     0.480000     0.165000      1.153250         0.50200   
max       0.815000     0.650000     1.130000      2.825500         1.48800   

       Viscera weight  Shell weigth        Rings  
count     4176.000000   4176.000000  4176.000000  
mean         0.180613      0.238852     9.932471  
std          0.109620      0.139213     3.223601  
min          0.000500      0.001500     1.000000  
25%          0.093375      0.130000     8.000000  
50%          0.171000      0.234000     9.000000  
75%          0.253000      0.329000    11.000000  
max          0.760000      1.005000    29.000000

In [114]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4176 entries, 0 to 4175
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Sex             4176 non-null   object 
 1   Length          4176 non-null   float64
 2   Diameter        4176 non-null   float64
 3   Height          4176 non-null   float64
 4   Whole weight    4176 non-null   float64
 5   Shucked weigth  4176 non-null   float64
 6   Viscera weight  4176 non-null   float64
 7   Shell weigth    4176 non-null   float64
 8   Rings           4176 non-null   int64  
dtypes: float64(7), int64(1), object(1)
memory usage: 293.8+ KB


In [115]:
df['Sex'].unique()

array(['M', 'F', 'I'], dtype=object)

In [116]:
df.dropna()
df

Sex  Length  Diameter  Height  Whole weight  Shucked weigth  \
0      M   0.350     0.265   0.090        0.2255          0.0995   
1      F   0.530     0.420   0.135        0.6770          0.2565   
2      M   0.440     0.365   0.125        0.5160          0.2155   
3      I   0.330     0.255   0.080        0.2050          0.0895   
4      I   0.425     0.300   0.095        0.3515          0.1410   
...   ..     ...       ...     ...           ...             ...   
4171   F   0.565     0.450   0.165        0.8870          0.3700   
4172   M   0.590     0.440   0.135        0.9660          0.4390   
4173   M   0.600     0.475   0.205        1.1760          0.5255   
4174   F   0.625     0.485   0.150        1.0945          0.5310   
4175   M   0.710     0.555   0.195        1.9485          0.9455   

      Viscera weight  Shell weigth  Rings  
0             0.0485        0.0700      7  
1             0.1415        0.2100      9  
2             0.1140        0.1550     10  
3             0.0395        0.0550      7  
4             0.0775        0.1200      8  
...              ...           ...    ...  
4171          0.2390        0.2490     11  
4172          0.2145        0.2605     10  
4173          0.2875        0.3080      9  
4174          0.2610        0.2960     10  
4175          0.3765        0.4950     12  

[4176 rows x 9 columns]

In [117]:
df.corr()

Length  Diameter    Height  Whole weight  Shucked weigth  \
Length          1.000000  0.986813  0.827552      0.925255        0.897905   
Diameter        0.986813  1.000000  0.833705      0.925452        0.893159   
Height          0.827552  0.833705  1.000000      0.819209        0.774957   
Whole weight    0.925255  0.925452  0.819209      1.000000        0.969403   
Shucked weigth  0.897905  0.893159  0.774957      0.969403        1.000000   
Viscera weight  0.903010  0.899726  0.798293      0.966372        0.931956   
Shell weigth    0.897697  0.905328  0.817326      0.955351        0.882606   
Rings           0.557123  0.575005  0.558109      0.540818        0.421256   

                Viscera weight  Shell weigth     Rings  
Length                0.903010      0.897697  0.557123  
Diameter              0.899726      0.905328  0.575005  
Height                0.798293      0.817326  0.558109  
Whole weight          0.966372      0.955351  0.540818  
Shucked weigth        0.931956      0.882606  0.421256  
Viscera weight        1.000000      0.907647  0.504274  
Shell weigth          0.907647      1.000000  0.628031  
Rings                 0.504274      0.628031  1.000000

#df.drop(labels='Rings', axis=1, inplace=True)
from sklearn.preprocessing import LabelEncoder
LE = LabelEncoder()
df['Sex_modified'] = LE.fit_transform(df['Sex'])
df.drop(labels = 'Sex', axis=1, inplace=True)
df

In [118]:
df_onehot= pd.get_dummies(df['Sex'])
#df= df.drop('Sex', axis=1, inplace=True)
df = df.join(df_onehot)
df


Sex  Length  Diameter  Height  Whole weight  Shucked weigth  \
0      M   0.350     0.265   0.090        0.2255          0.0995   
1      F   0.530     0.420   0.135        0.6770          0.2565   
2      M   0.440     0.365   0.125        0.5160          0.2155   
3      I   0.330     0.255   0.080        0.2050          0.0895   
4      I   0.425     0.300   0.095        0.3515          0.1410   
...   ..     ...       ...     ...           ...             ...   
4171   F   0.565     0.450   0.165        0.8870          0.3700   
4172   M   0.590     0.440   0.135        0.9660          0.4390   
4173   M   0.600     0.475   0.205        1.1760          0.5255   
4174   F   0.625     0.485   0.150        1.0945          0.5310   
4175   M   0.710     0.555   0.195        1.9485          0.9455   

      Viscera weight  Shell weigth  Rings  F  I  M  
0             0.0485        0.0700      7  0  0  1  
1             0.1415        0.2100      9  1  0  0  
2             0.1140        0.1550     10  0  0  1  
3             0.0395        0.0550      7  0  1  0  
4             0.0775        0.1200      8  0  1  0  
...              ...           ...    ... .. .. ..  
4171          0.2390        0.2490     11  1  0  0  
4172          0.2145        0.2605     10  0  0  1  
4173          0.2875        0.3080      9  0  0  1  
4174          0.2610        0.2960     10  1  0  0  
4175          0.3765        0.4950     12  0  0  1  

[4176 rows x 12 columns]

In [119]:
#Height column with null values
(df["Height"]==0).sum()

2

In [120]:
df[df["Height"]==0]

Sex  Length  Diameter  Height  Whole weight  Shucked weigth  \
1256   I   0.430      0.34     0.0         0.428          0.2065   
3995   I   0.315      0.23     0.0         0.134          0.0575   

      Viscera weight  Shell weigth  Rings  F  I  M  
1256          0.0860        0.1150      8  0  1  0  
3995          0.0285        0.3505      6  0  1  0

In [121]:
means_sex = pd.pivot_table(df, index=['Sex'], aggfunc = {'Height':np.mean})
means_sex

Height
Sex          
F    0.158011
I    0.107996
M    0.151418

In [122]:
df["Height"]=df["Height"].replace(to_replace=0, value=0.1077996)

In [123]:
df['Rings_year'] = df['Rings']+1.5
df['Rings_year']

0        8.5
1       10.5
2       11.5
3        8.5
4        9.5
        ... 
4171    12.5
4172    11.5
4173    10.5
4174    11.5
4175    13.5
Name: Rings_year, Length: 4176, dtype: float64

In [124]:
df.drop(labels='Rings', axis=1)

Sex  Length  Diameter  Height  Whole weight  Shucked weigth  \
0      M   0.350     0.265   0.090        0.2255          0.0995   
1      F   0.530     0.420   0.135        0.6770          0.2565   
2      M   0.440     0.365   0.125        0.5160          0.2155   
3      I   0.330     0.255   0.080        0.2050          0.0895   
4      I   0.425     0.300   0.095        0.3515          0.1410   
...   ..     ...       ...     ...           ...             ...   
4171   F   0.565     0.450   0.165        0.8870          0.3700   
4172   M   0.590     0.440   0.135        0.9660          0.4390   
4173   M   0.600     0.475   0.205        1.1760          0.5255   
4174   F   0.625     0.485   0.150        1.0945          0.5310   
4175   M   0.710     0.555   0.195        1.9485          0.9455   

      Viscera weight  Shell weigth  F  I  M  Rings_year  
0             0.0485        0.0700  0  0  1         8.5  
1             0.1415        0.2100  1  0  0        10.5  
2             0.1140        0.1550  0  0  1        11.5  
3             0.0395        0.0550  0  1  0         8.5  
4             0.0775        0.1200  0  1  0         9.5  
...              ...           ... .. .. ..         ...  
4171          0.2390        0.2490  1  0  0        12.5  
4172          0.2145        0.2605  0  0  1        11.5  
4173          0.2875        0.3080  0  0  1        10.5  
4174          0.2610        0.2960  1  0  0        11.5  
4175          0.3765        0.4950  0  0  1        13.5  

[4176 rows x 12 columns]

In [125]:
#rearranging columns
df = df[['Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weigth', 'Viscera weight', 'Shell weigth', 'F', 'I', 'M', 'Rings_year']]
cols = list(df.columns.values)
cols
df

Length  Diameter  Height  Whole weight  Shucked weigth  Viscera weight  \
0      0.350     0.265   0.090        0.2255          0.0995          0.0485   
1      0.530     0.420   0.135        0.6770          0.2565          0.1415   
2      0.440     0.365   0.125        0.5160          0.2155          0.1140   
3      0.330     0.255   0.080        0.2050          0.0895          0.0395   
4      0.425     0.300   0.095        0.3515          0.1410          0.0775   
...      ...       ...     ...           ...             ...             ...   
4171   0.565     0.450   0.165        0.8870          0.3700          0.2390   
4172   0.590     0.440   0.135        0.9660          0.4390          0.2145   
4173   0.600     0.475   0.205        1.1760          0.5255          0.2875   
4174   0.625     0.485   0.150        1.0945          0.5310          0.2610   
4175   0.710     0.555   0.195        1.9485          0.9455          0.3765   

      Shell weigth  F  I  M  Rings_year  
0           0.0700  0  0  1         8.5  
1           0.2100  1  0  0        10.5  
2           0.1550  0  0  1        11.5  
3           0.0550  0  1  0         8.5  
4           0.1200  0  1  0         9.5  
...            ... .. .. ..         ...  
4171        0.2490  1  0  0        12.5  
4172        0.2605  0  0  1        11.5  
4173        0.3080  0  0  1        10.5  
4174        0.2960  1  0  0        11.5  
4175        0.4950  0  0  1        13.5  

[4176 rows x 11 columns]

In [126]:
#Fill in these null values with mean. these values are for infants



In [127]:
df['Rings_year'].min() #2.5
df['Rings_year'].max() #30.5


30.5

In [128]:
df["Rings_year"].value_counts().sort_index()

2.5       1
3.5       1
4.5      15
5.5      57
6.5     115
7.5     259
8.5     391
9.5     568
10.5    689
11.5    634
12.5    487
13.5    267
14.5    203
15.5    126
16.5    102
17.5     67
18.5     58
19.5     42
20.5     32
21.5     26
22.5     14
23.5      6
24.5      9
25.5      2
26.5      1
27.5      1
28.5      2
30.5      1
Name: Rings_year, dtype: int64

In [129]:
bin_size = len(df['Rings_year'])
bin_size

4176

In [152]:
def cube_root(x):
    return x**(1/3)
my_bin = print(cube_root(bin_size)*2)

32.20699152479751


In [174]:
df_list = df.index
df["Rings"] = ''
for i in df_list:
    if df["Rings_year"][i]<=12:
        df["Rings"].loc[i]=0
    elif 12<df['Rings_year'][i]<=22:
        df['Rings'].loc[i]=1
    else:
        df['Rings'].loc[i]=2
df['Rings']=df["Rings"].astype(int)
df
#['Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weigth', 'Viscera weight', 'Shell weigth', 'F', 'I', 'M', 'Rings_year']

<ipython-input-174-371a15869544>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Rings"] = ''
/Users/trimpu/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)
<ipython-input-174-371a15869544>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Rings"].loc[i]=0
<ipyth

Length  Diameter  Height  Whole weight  Shucked weigth  Viscera weight  \
0      0.350     0.265   0.090        0.2255          0.0995          0.0485   
1      0.530     0.420   0.135        0.6770          0.2565          0.1415   
2      0.440     0.365   0.125        0.5160          0.2155          0.1140   
3      0.330     0.255   0.080        0.2050          0.0895          0.0395   
4      0.425     0.300   0.095        0.3515          0.1410          0.0775   
...      ...       ...     ...           ...             ...             ...   
4171   0.565     0.450   0.165        0.8870          0.3700          0.2390   
4172   0.590     0.440   0.135        0.9660          0.4390          0.2145   
4173   0.600     0.475   0.205        1.1760          0.5255          0.2875   
4174   0.625     0.485   0.150        1.0945          0.5310          0.2610   
4175   0.710     0.555   0.195        1.9485          0.9455          0.3765   

      Shell weigth  F  I  M  Rings_year  Rings  
0           0.0700  0  0  1         8.5      0  
1           0.2100  1  0  0        10.5      0  
2           0.1550  0  0  1        11.5      0  
3           0.0550  0  1  0         8.5      0  
4           0.1200  0  1  0         9.5      0  
...            ... .. .. ..         ...    ...  
4171        0.2490  1  0  0        12.5      1  
4172        0.2605  0  0  1        11.5      0  
4173        0.3080  0  0  1        10.5      0  
4174        0.2960  1  0  0        11.5      0  
4175        0.4950  0  0  1        13.5      1  

[4176 rows x 12 columns]

In [175]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4176 entries, 0 to 4175
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Length          4176 non-null   float64
 1   Diameter        4176 non-null   float64
 2   Height          4176 non-null   float64
 3   Whole weight    4176 non-null   float64
 4   Shucked weigth  4176 non-null   float64
 5   Viscera weight  4176 non-null   float64
 6   Shell weigth    4176 non-null   float64
 7   F               4176 non-null   uint8  
 8   I               4176 non-null   uint8  
 9   M               4176 non-null   uint8  
 10  Rings_year      4176 non-null   float64
 11  Rings           4176 non-null   int64  
dtypes: float64(8), int64(1), uint8(3)
memory usage: 306.0 KB


In [176]:
#Neural Network - Keras

In [177]:
from keras.layers import Dense
from keras.models import Sequential
from sklearn import preprocessing
#from numpy import unique
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

dataset = df.values
# split into input (X) and output (y) variables

X= df.drop(["Rings_year", 'Rings'], axis=1)
y = df["Rings"]
print("x", X.shape)
print("y", y.shape)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

model = Sequential()
model.add(Dense(20, activation = 'relu', input_shape=(10, )))
model.add(Dense(10, activation='relu'))
model.add(Dense(3, activation="softmax"))
model.compile(optimizer ='adam', loss="sparse_categorical_crossentropy", metrics=["accuracy"])
model.fit(X_train, y_train, epochs=100, verbose=0, batch_size=20)

y_pred = model.predict(X_test)
y_pred_list =[]

for each in y_pred:
    y = y_pred_list.append(np.argmax(y))
    
print(classification_report(y_test, y_pred_list))
print(accuracy_score(y_test, y_pred_list))

x (4176, 10)
y (4176,)
              precision    recall  f1-score   support

           0       0.66      1.00      0.80       829
           1       0.00      0.00      0.00       411
           2       0.00      0.00      0.00        13
          82       0.00      0.00      0.00         0

    accuracy                           0.66      1253
   macro avg       0.17      0.25      0.20      1253
weighted avg       0.44      0.66      0.53      1253

0.6608140462889066


/Users/trimpu/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/trimpu/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/trimpu/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/trimpu/opt/anacon

4. Write another algorithm to predict the same result as the previous question using either KNN or logistic regression.

In [49]:
from sklearn import preprocessing
from sklearn.linear_model import LogisticRegression
#from sklearn import StandardScalar

from sklearn import utils

target=df['Rings_year']
feature=df.drop(['Rings_year'],axis=1)
X = feature.values
y = target.values

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)
#lr = LogisticRegression()
#lr.fit(X_train, y_train)

lab_enc = preprocessing.LabelEncoder()
training_scores_encoded = lab_enc.fit_transform(y_train)
print(training_scores_encoded)
print(utils.multiclass.type_of_target(y_train))
print(utils.multiclass.type_of_target(y_train.astype('int')))
print(utils.multiclass.type_of_target(training_scores_encoded))

lr = LogisticRegression()
lr.fit(X_train, training_scores_encoded)
print("LogisticRegression")
print(lr.score(X_train, training_scores_encoded))

#X = preprocessing.normalize(x)

[11  6 11 ...  4  6  0]
continuous
multiclass
multiclass
LogisticRegression
0.2595785440613027


/Users/trimpu/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


5. Create a neural network using pytorch to predict the same result as question 3. 

In [132]:
import pandas as pd
import torch

In [138]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X = torch.from_numpy(df.drop(['Rings', "Rings_year"],axis = 1).to_numpy())
y= df['Rings'].values

#X_train = torch.from_numpy(df.drop('class', axis=1).to_numpy())

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

sc=StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [139]:
import torch.nn as nn
import torch.nn.functional as F # this has activation functions

#Creating tensors
X_train=torch.FloatTensor(X_train)
X_test = torch.FloatTensor(X_test)

y_train = torch.LongTensor(y_train)
y_test = torch.LongTensor(y_test)

print(X_train)

tensor([[ 0.2138,  0.3745, -0.3526,  ..., -0.6795,  1.4476, -0.7518],
        [-0.4146, -0.2852, -0.4709,  ..., -0.6795,  1.4476, -0.7518],
        [-0.4146, -0.5896, -0.5892,  ...,  1.4717, -0.6908, -0.7518],
        ...,
        [-0.0376,  0.2222,  0.2390,  ...,  1.4717, -0.6908, -0.7518],
        [ 0.4233,  0.5267,  0.2390,  ...,  1.4717, -0.6908, -0.7518],
        [ 0.5908,  0.5774,  0.0024,  ...,  1.4717, -0.6908, -0.7518]])


In [140]:
class ANN_Model(nn.Module):
    def __init__(self, input_features=10, hidden1=20, hidden2=20, out_features=3):
        super().__init__()
        self.layer_1_connection = nn.Linear(input_features, hidden1)
        self.layer_2_connection = nn.Linear(hidden1, hidden2)
        self.out = nn.Linear(hidden2, out_features)
    
    def forward(self, x):
        #apply activation functions
        x = F.relu(self.layer_1_connection(x))
        x = F.relu(self.layer_2_connection(x))
        x = self.out(x)
        return x

In [141]:
torch.manual_seed(42)

#instantiate the model
model = ANN_Model()

In [142]:
# Loss function
loss_function = nn.CrossEntropyLoss() # we us ethis when output is 1D array

#optimizer
optimizer = torch.optim.Adam(model.parameters(), lr=0.01)

In [143]:
final_loss=[]
n_epochs = 500
for epoch in range(n_epochs):
    y_pred = model.forward(X_train)
    loss = loss_function(y_pred, y_train)
    final_loss.append(loss)
    
    if epoch % 10 ==1:
        print(f'Epoch number: {epoch} with loss: {loss.item()}')
        
    optimizer.zero_grad() #zero the gradient before running backwards
    loss.backward() #for backward propagation
    optimizer.step() #performs one optimization step each epoch

Epoch number: 1 with loss: 1.0444133281707764
Epoch number: 11 with loss: 0.5567077994346619
Epoch number: 21 with loss: 0.5260249972343445
Epoch number: 31 with loss: 0.4942449629306793
Epoch number: 41 with loss: 0.4780687093734741
Epoch number: 51 with loss: 0.4697984755039215
Epoch number: 61 with loss: 0.46135571599006653
Epoch number: 71 with loss: 0.4495050311088562
Epoch number: 81 with loss: 0.43314969539642334
Epoch number: 91 with loss: 0.4204024076461792
Epoch number: 101 with loss: 0.4117338955402374
Epoch number: 111 with loss: 0.405229777097702
Epoch number: 121 with loss: 0.40193066000938416
Epoch number: 131 with loss: 0.39935103058815
Epoch number: 141 with loss: 0.39711251854896545
Epoch number: 151 with loss: 0.39519986510276794
Epoch number: 161 with loss: 0.3935389220714569
Epoch number: 171 with loss: 0.3924194276332855
Epoch number: 181 with loss: 0.3911898136138916
Epoch number: 191 with loss: 0.38973256945610046
Epoch number: 201 with loss: 0.3884645402431488


In [144]:
#predictions
y_pred = []

with torch.no_grad():
    for i, data in enumerate(X_test):
        prediction = model(data)
        y_pred.append(prediction.argmax().item())

In [145]:
from sklearn.metrics import accuracy_score
a_score = accuracy_score(y_test, y_pred)
a_score

0.8145933014354066

In [146]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.78      0.72      0.75       281
           1       0.83      0.89      0.86       537
           2       0.00      0.00      0.00        18

    accuracy                           0.81       836
   macro avg       0.54      0.54      0.54       836
weighted avg       0.80      0.81      0.80       836



6. Compare the performance of the neural networks to the other model you created. Which performed better? Why do you think that is?

Neural networks performed better than Logistic and KNN. Pytorch with 81% accuracy, Keras with 66% and LogisticRegression with 25%. 
This is because of two things:
1. Neural networks are capable of extreme interactions with all the features
2. These interactions can be increased to get more accurate results, by increasing nodes/layers. Thus increases interactions.

